enhancements/multi-arch/dynamic-imagestream-importmode-setting.md

```diff
@@ -0,0 +1,408 @@
+
+---
+title: dynamic-imagestream-importmode-setting
+authors:
+  - "@Prashanth684"
+reviewers:
+  - "@deads2k, apiserver"
+  - "@JoelSpeed, API"
+  - "@wking, CVO"
+  - "@dmage, Imagestreams"
+  - "@soltysh, Workloads, oc"
+approvers:
+  - "@deads2k"
+api-approvers:
+  - "@JoelSpeed"
+creation-date: 2024-04-02
+last-updated: 2024-04-02
+tracking-link:
+  - https://issues.redhat.com/browse/MULTIARCH-4552
+---
+
+# Dynamically set Imagestream importMode based on cluster payload type
+
+## Summary
+
+This is a proposal to set ImportMode for
+imagestreams dynamically based on the payload type
+of the cluster. This means that if the OCP release
+payload is a singlemanifest payload, the
+importMode for all imagestreams by default will be
+`Legacy` and if the payload is a multi payload the
+default will be `PreserveOriginal`. 
+
+## Motivation
+
+This change ensures that imagestreams will be
+compatible for single and multi arch compute
+clusters without needing manual intervention.
+Users who install or upgrade to the multi
+release payload
+(https://mirror.openshift.com/pub/openshift-v4/multi/)
+and want to add compute nodes of differing
+architectures, need not change the behaviour
+of any newly imported imagestreams or any scripts
+which import them. For a cluster installed with
+single arch payload, imagestreams default import
+mode would always be set to `Legacy` because it
+is sufficient if the single image manifest is
+imported and for a cluster installed with the
+multi payload it would be `PreserveOriginal`
```

Comment:
Pinning down current behavior, [multi-arch release controller][1] -> [recent 4.16 nightly][2] -> [blocker run][3] -> [gathered artifacts][4] -> ... -> [must-gather artifacts][5]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-multiarch-master-nightly-4.16-ocp-e2e-aws-ovn-heterogeneous/1782426110521249792/artifacts/ocp-e2e-aws-ovn-heterogeneous/gather-must-gather/artifacts/must-gather.tar | tar -tvz | grep 'namespaces/openshift/.*imagestreams/installer.yaml'
-rw------- 1017300000/root    2306 2024-04-22 09:47 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-1e4b8e008b76da76e5b9f893667d39d627f442cc2dfa4139e0084dc679786828/namespaces/openshift/image.openshift.io/imagestreams/installer.yaml
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-multiarch-master-nightly-4.16-ocp-e2e-aws-ovn-heterogeneous/1782426110521249792/artifacts/ocp-e2e-aws-ovn-heterogeneous/gather-must-gather/artifacts/must-gather.tar | tar -xOz quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-1e4b8e008b76da76e5b9f893667d39d627f442cc2dfa4139e0084dc679786828/namespaces/openshift/image.openshift.io/imagestreams/installer.yaml | yaml2json | jq -c '.spec.tags[].importPolicy'
{"importMode":"PreserveOriginal","scheduled":true}
```

And [amd64 release controller][6] -> [4.16.0-ec.5][7] -> [blocker run][8] -> [gathered artifacts][9] -> ... -> [must-gather artifacts][10]:

```console
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1777335045871112192/artifacts/e2e-aws-ovn-serial/gather-must-gather/artifacts/must-gather.tar | tar -tvz | grep 'namespaces/openshift/.*imagestreams/installer.yaml'
-rw------- 1010890000/root     2306 2024-04-08 09:46 quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-6496aec41107e9730791d82ebc65bb873aa228c0fd873444932b477679feb1ff/namespaces/openshift/image.openshift.io/imagestreams/installer.yaml
$ curl -s https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1777335045871112192/artifacts/e2e-aws-ovn-serial/gather-must-gather/artifacts/must-gather.tar | tar -xOz quay-io-openshift-release-dev-ocp-v4-0-art-dev-sha256-6496aec41107e9730791d82ebc65bb873aa228c0fd873444932b477679feb1ff/namespaces/openshift/image.openshift.io/imagestreams/installer.yaml | yaml2json | jq -c '.spec.tags[].importPolicy'
{"importMode":"PreserveOriginal","scheduled":true}
```

What's the downside to that approach?  I could see mirror-size concerns, but you list sparse-manifests as a non-goal, and while I could see not needing a particular image in all architectures, I'm having trouble seeing cluster-scoped decisions in that space.  And... by the time you're launching clusters off of a multi-arch mirror, you've already completed the mirroring, right?  Without sparse-manifests, how do you even install a multi-arch payload based on partial mirroring?

[1]: https://multi.ocp.releases.ci.openshift.org/
[2]: https://multi.ocp.releases.ci.openshift.org/releasestream/4.16.0-0.nightly-multi/release/4.16.0-0.nightly-multi-2024-04-22-110357
[3]: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-multiarch-master-nightly-4.16-ocp-e2e-aws-ovn-heterogeneous/1782426110521249792
[4]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-multiarch-master-nightly-4.16-ocp-e2e-aws-ovn-heterogeneous/1782426110521249792/
[5]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-multiarch-master-nightly-4.16-ocp-e2e-aws-ovn-heterogeneous/1782426110521249792/artifacts/ocp-e2e-aws-ovn-heterogeneous/gather-must-gather/artifacts/
[6]: https://amd64.ocp.releases.ci.openshift.org/
[7]: https://amd64.ocp.releases.ci.openshift.org/releasestream/4-dev-preview/release/4.16.0-ec.5
[8]: https://prow.ci.openshift.org/view/gs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1777335045871112192
[9]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1777335045871112192/
[10]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/logs/periodic-ci-openshift-release-master-nightly-4.16-e2e-aws-ovn-serial/1777335045871112192/artifacts/e2e-aws-ovn-serial/gather-must-gather/artifacts/

Reply:
Looks like those OpenShift ImageStreams moved to explicitly `PreserveOriginal` in openshift/cluster-samples-operator#482, which seems to have gone out in 4.13.

Reply:
This proposal is meant to set `importMode` default based on the type of payload. Today, the default is `Legacy` which is problematic when running on a cluster with multiarch compute nodes as the imported image is manifestlisted. This is not a problem for CVO managed imagestreams as we default them to `preserveOriginal`, but for user managed imagestreams, the default is set through apiserver. Also, sparse manifest in this context refers to sparse mirroring where a user wants to mirror only certain arches, but keep the payload digest intact. this is not supported by any registry today, but it would help users who are concerned with mirroring all 4  arches in any of their images.

Reply:
Yup, [RFE-3333](https://issues.redhat.com/browse/RFE-3333) and [MULTIARCH-424](https://issues.redhat.com/browse/MULTIARCH-4247) seem like reasonable public references for sparse-manifest mirroring, and while that seems extremely useful, I understand that it's a larger lift than you're aiming for in this enhancement.

If the target of this enhancement is to create an even-more-clever default that works well for more ImageStream creators without requiring an explicit `importMode`, what's keeping us from adding additional steps to [the `Legacy` logic][1] like "don't flatten to a single arch if the cluster has nodes from multiple arches"?

[1]: https://github.com/openshift/api/blob/a8b3ba923728bafe72b96dbfc0e4cfbba769d226/image/v1/types.go#L321

Reply:
i don't think we need to build this level of intelligence into apiserver/image registry operator. eventually, when sparse manifests are supported, importing manifestlists as default will become a non issue, so we want to keep it simple.